File: testdata/fixtures/java/md5-used-as-password.java
import java.security.MessageDigest;

public class Md5Password {
    Store store;

    public void vuln(String pw) throws Exception {
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] digest = md.digest(pw.getBytes());
        // ruleid: md5-used-as-password
        store.savePassword(digest);
    }

    public void vulnAgain(String pw) throws Exception {
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] hash = md.digest(pw.getBytes());
        // ruleid: md5-used-as-password
        store.savePassword(hash);
    }

    public void vulnThird(String pw) throws Exception {
        MessageDigest md = MessageDigest.getInstance("MD5");
        byte[] sum = md.digest(pw.getBytes());
        // ruleid: md5-used-as-password
        store.updatePassword(sum);
    }

    public void safe(String pw) throws Exception {
        // ok: md5-used-as-password
        store.savePassword("changeme");
    }

    public void safeAgain(String pw) {
        // ok: md5-used-as-password
        store.savePassword(pw);
    }

    public void plain(String pw) {
        // ok: md5-used-as-password
        System.out.println(pw);
    }
}

File: testdata/fixtures/go/md5-used-as-password.go
package fixture

import (
	"crypto/md5"
	"crypto/sha256"
	"fmt"
)

func vulnNew(pw string) string {
	h := md5.New()
	h.Write([]byte(pw))
	hex := fmt.Sprintf("%x", h.Sum(nil))
	// ruleid: md5-used-as-password
	return savePassword(hex)
}

func vulnSum(pw string) string {
	sum := md5.Sum([]byte(pw))
	hex := fmt.Sprintf("%x", sum)
	// ruleid: md5-used-as-password
	return updatePassword(hex)
}

func vulnDirect(pw string) string {
	h := md5.New()
	h.Write([]byte(pw))
	// ruleid: md5-used-as-password
	return savePassword(fmt.Sprintf("%x", h.Sum(nil)))
}

func safeSha256(pw string) string {
	h := sha256.New()
	h.Write([]byte(pw))
	hex := fmt.Sprintf("%x", h.Sum(nil))
	// ok: md5-used-as-password
	return savePassword(hex)
}

func safeNoPassword(data []byte) string {
	h := md5.New()
	h.Write(data)
	// ok: md5-used-as-password
	return fmt.Sprintf("%x", h.Sum(nil))
}

func safeConstant() string {
	// ok: md5-used-as-password
	return savePassword("changeme")
}
